apps/browser/lib/services/evo/EvoThread.ts

```diff
@@ -0,0 +1,249 @@
+    this._callbacks.setWorkspace(this._state.workspace);
+  }
+
+  async start(options: EvoThreadStartOptions) {
+    const {
+      goal,
+      allowTelemetry,
+      openAiApiKey
+    } = options;
+
+    // Wait until loading has finished
+    await this.waitForLoad();
+
+    if (this._state.isRunning) {
+      this._callbacks?.onError("A goal is already underway.");
+      return;
+    }
+
+    this.setIsRunning(true);
+
+    // Acquire a GoalID
+    const subsidize = !openAiApiKey;
+    const goalId = await GoalApi.create(
+      this.chatId,
+      allowTelemetry ? goal : "<redacted>",
+      subsidize,
+      () => this._callbacks?.onGoalCapReached()
+    );
+
+    if (!goalId) {
+      this._callbacks?.onError("Unable to acquire a goal ID.");
+      this.setIsRunning(false);
+      return;
+    }
+
+    // Create an Evo instance
+    const evo = createEvoInstance(
+      goalId,
+      this._state.workspace,
+      options.openAiApiKey,
+      this._config.onMessagesAdded,
+      this._config.onVariableSet,
+      (chatLog) =>
+        this.onChatLog(chatLog) || Promise.resolve(),
+      () =>
+        this._callbacks?.onGoalCapReached(),
+      // onError
+      (error) =>
+        this._callbacks?.onError(error)
+    );
+
+    if (!evo) {
+      this.setIsRunning(false);
+      return;
+    }
+
+    // Run the evo instance against the goal
+    await this.runEvo(evo, options.goal);
+  }
+
+  private async load() {
+    const chatId = this._config.chatId;
+    this._state.isLoading = true;
+
+    const results = await Promise.all<[
+      Promise<ChatLog[]>,
+      Promise<Workspace>
+    ]>([
+      this._config.loadChatLog(chatId).catch((reason) => {
+        this._callbacks?.onError(reason.toString());
+        return [];
+      }),
+      this._config.loadWorkspace(chatId).catch((reason) => {
+        this._callbacks?.onError(reason.toString());
+        return new InMemoryWorkspace();
+      })
+    ]);
+
+    this._state.logs = results[0];
+    this._state.workspace = results[1];
+    this._state.isLoading = false;
+  }
+
+  private async waitForLoad() {
+    while (this._state.isLoading) {
+      await new Promise((resolve) =>
+        setTimeout(resolve, 200)
+      );
+    }
+    return Promise.resolve();
+  }
+
+  private setIsRunning(value: boolean) {
+    this._state.isRunning = value;
+    this._callbacks?.setIsRunning(value);
+  }
+
+  private async onChatLog(chatLog: ChatLog): Promise<void> {
+    this._state.logs = [...this._state.logs, chatLog];
+    this._callbacks?.setChatLog(this._state.logs);
+    await this._config.onChatLogAdded(chatLog);
+  }
+
+  private async runEvo(evo: Evo, goal: string): Promise<void> {
+    const iterator = evo.run({ goal });
+
+    await this.onChatLog({
+      title: goal,
+      user: "user"
+    });
+
+    let stepCounter = 1;
+
+    while (this._state.isRunning) {
+      const response = await iterator.next();
+
+      this._callbacks?.setWorkspace(this._state.workspace);
```

Comment:
what's the reason of this line? 

the example prompt of the CSV doesn't work with it; or if you try to add a file before running Evo and then run a prompt the workspace is restarted

Reply:
To solve this, we'd want to create a ChatID when a file is uploaded to the workspace. This however is a bit of a touchy state update issue so I'd rather not do this in this PR, and do it after this & the redesign PR are merged. Great find though, thank you!

Reply:
https://github.com/polywrap/evo.ninja/issues/594